Feature/select files zip 91768006

AurL opened the pull request:
Craft archive are more accurately built :
- Only useful files are added
- Converts (by default) MBM into PNGs (option to disable it)
- Height maps to normal maps conversion

AurL committed:
[#91768006] Selects more efficiently the files to pack

At the beginning, we retrieved all the files that were in the directory of
the part.cfg (part description file) of the parts that were used in the
current craft. To save space (especially for the cases where a directory
contains several parts), we check deeper in the files to retrieve
only the files that are actually used (.cfg + .mu + .mbm(s)).

AurL committed:
[#91768005] Adds MBM to PNG conversion

MBM is the format used by KSP for models'textures. MBM is a derivative of
PBM (Portable Bit Map) which is heavy and makes the craft archives to be
huge. Texture conversion (MBM -> PNG) is now done by the exporter to
reduce final archive size, which allows the user to be under the size
limit if not PRO.

AurL committed:
[#91768704] Adds heightmap to normal map conversion

Since the data to identify which textures need to
be converted to normal maps is in the MBM header, we also perform the
heightmap to normal map conversion.

AurL committed:
Adds README

AurL committed:
[#91768006] Fixes heightmap to normal map conversion

cedricpinson commented on texture_converter.py:
```diff
@@ -0,0 +1,69 @@
+            for x in range(1, (width * nb_components) -1, nb_components):
+                norm_val = QtGui.QVector3D(pixels[y][x+1] - pixels[y][x-1], pixels[y+1][x] - pixels[y-1][x], 255)
+                norm_val.normalize()
+                norm_val = norm_val.__mul__(QtGui.QVector3D(127, 127, 127))
```
it means it can't work from command line right ? you need to have qt4 installed. If it's the only deps maybe we can add some math func in the code instead of using qt4

AurL replied:
Yes, you are right. I used Qt4 because I needed some Vector features for the previous version of the map conversion, but with this version it is easier to implement them. To make it ok for command line, I will write these methods or find a more convenient way to do this.

AurL committed:
[#91768006] Makes py files flake8 compliant

AurL committed:
[#91768006] Improves textures conversion

The conversion now uses PIL functions to convert more efficiently
textures from MBM to PNG, also handling heightmap to normal map conversion.

AurL committed:
[#92703668] Removes temp files after upload

AurL committed:
[#91768006] Fixes/improves exporter GUI

AurL committed:
[#92821496] Replaces 'title' field by 'name' for skfb API

AurL committed:
[#92703668] Makes temp file deletion safer

AurL committed:
Adds Windows build beta

AurL committed:
Adds ksp2skfb icon

AurL committed:
Updates README

marchelbling commented on gui.py:
```diff
@@ -12,99 +14,169 @@ def __init__(self, parent=None):
+                    need_clear = True
+                    print(os.getcwd())
+            if need_clear:
+                os.remove('userdata.json')
```
why the need for an extra variable? the only code path that leads to this is when an exception is raised; move the `os.remove` in the exceptino block. The `print` statement is likely a debug stuff; if so remove it

marchelbling commented on gui.py:
```diff
@@ -12,99 +14,169 @@ def __init__(self, parent=None):
+
+    def description_input_limit(self):
+        if len(self.description_tb.toPlainText()) > 1024:
+            self.description_tb.setText(self.description_tb.toPlainText()[:1024])
```
python silcing is developer friendly:

```
>>> 'foo'[:1000]
'foo'
```

No need to test the length explicitely

marchelbling commented on gui.py:
```diff
@@ -12,99 +14,169 @@ def __init__(self, parent=None):
+    def description_input_limit(self):
+        if len(self.description_tb.toPlainText()) > 1024:
+            self.description_tb.setText(self.description_tb.toPlainText()[:1024])
+        self.description_label.setText('Description ({}/1024)'.format(len(self.description_tb.toPlainText())))
```
not sure the `({}/1024)` is totally understandable from a user point of view. Maybe say 1024 characters?

marchelbling commented on gui.py:
```diff
@@ -12,99 +14,169 @@ def __init__(self, parent=None):
-            self.manager = KSP2Skfb(self.game_dir)
-            self.craft_list = self.manager.get_craft_list()
+        self.enable_skfb_ui(True)
+        self.name_tb.setText('{}'.format(self.craft_list[index].split('.')[1].strip()))
```
it's difficult to understand what `self.craft_list[index].split('.')[1].strip())` does here; would probably be better in a rightly named function

marchelbling commented on gui.py:
```diff
@@ -113,51 +185,62 @@ def update_status(self, status):
                     QtGui.QDesktopServices.openUrl(QtCore.QUrl(url))
+
+                    # Save the sketchfab api token in user_data
+                    self.user_data['api'] = '{}'.format(self.api_token_tb.text())
```
`format` call probably unnecessary

marchelbling commented on gui.py:
```diff
@@ -166,44 +249,65 @@ def upload_progress(value, max):
+        self.tags_tb.setEnabled(bool_value)
 
+    def set_craft_list(self):
+        self.craft_list = []
```
the attribut is set two lines below so is this necessary?

marchelbling commented on gui.py:
```diff
@@ -166,44 +249,65 @@ def upload_progress(value, max):
+        self.game_dir = game_dir
+        self.game_path_tb.setText(self.game_dir)
+        if os.path.exists(game_dir):
+            self.manager.set_game_dir(str(game_dir))
```
shouldn't `str(game_dir)` be `to_utf8(game_dir)`?

marchelbling commented on kspmanager.py:
```diff
@@ -66,15 +68,16 @@ def part_parameter(key, value):
         modelPart.setHeader(QtNetwork.QNetworkRequest.ContentTypeHeader, "application/octet-stream")
-        modelPart.setHeader(QtNetwork.QNetworkRequest.ContentDispositionHeader, "form-data; name=\"modelFile\"; filename=\"%s\"" % (archive))
+        modelPart.setHeader(QtNetwork.QNetworkRequest.ContentDispositionHeader,
+                            "form-data; name=\"modelFile\"; filename=\"%s\"" % (archive))
```
=> `filename=\"{}\"".format(archive)`

marchelbling commented on kspmanager.py:
```diff
@@ -158,41 +206,91 @@ def get_craft_path(self, name):
+        ''' Get a set of craft assets '''
+        assets_set = set()
+        for node in craftnodes.nodes:
+                label, value = node[1].values[0]
```
looks like double indentation; beware

marchelbling commented on kspmanager.py:
```diff
@@ -158,41 +206,91 @@ def get_craft_path(self, name):
+                        os.remove(d)
+                    except:
+                        pass
+        print('Temp files cleaned')
```
a lot of `if`s, a lot of catch all exceptions. Is this really necessary? Tempfiles are meant to be deleted upon computer restart

marchelbling commented on mureader.py:
```diff
@@ -0,0 +1,679 @@
+    ET_MESH_RENDERER = 8
+    ET_SKINNED_MESH_RENDERER = 9
+    ET_MATERIALS = 10
+    ET_MATERIAL = 11    #XXX not used?
```
what's this `XXX` comment about?

AurL replied:
I don't know, I reused (naively) the code from io_mu for this part.

marchelbling commented on mureader.py:
```diff
@@ -0,0 +1,679 @@
+
+class MuTexture:
+    def __init__(self):
+        pass
```
useless `__init__`; please remove

marchelbling commented on mureader.py:
```diff
@@ -0,0 +1,679 @@
+        self.name = mu.read_string()
+        self.type = mu.read_int()
+        #print("   ", self.name, self.type)
+        return self
```
I don't like this style a lot (having an empty init and an instance method returning `self` after some initialization). It looks like `read` is the actual constructor.

marchelbling commented on mureader.py:
```diff
@@ -0,0 +1,679 @@
+        self.lbSize = mu.read_vector()
+        self.wrapMode = mu.read_int()
+        num_curves = mu.read_int()
+        for i in range(num_curves):
```
Use a list comprehension with silent variable
`self.curves = [MuCurve().read(mu) for _ in range(num_curves)]` (or with my previous remark `self.curves = [MuCurve(mu) for _ in range(num_curves)]`

marchelbling commented on mureader.py:
```diff
@@ -0,0 +1,679 @@
+
+class MuCollider_Base:
+    def __init__(self, type):
+        self.type = type
```
beware of name shadowing; `type` is a python [builtin](https://docs.python.org/2/library/functions.html#type)

AurL commented:
I am closing this PR since another one is up-to-date and includes all these changes.
